cli/build.gradle.kts

```diff
@@ -123,6 +123,7 @@ dependencies {
 
     implementation("com.fasterxml.jackson.module:jackson-module-kotlin:$jacksonVersion")
     implementation("com.github.ajalt.clikt:clikt:$cliktVersion")
+    implementation("io.github.config4k:config4k:$config4kVersion")
```

Comment:
Could there be an alternative to introducing another configuration library? Maybe just use Jackson to serialize the classes to JSON?

Reply:
> Could there be an alternative to introducing another configuration library? Maybe just use Jackson to serialize the classes to JSON?

The problem is that we need the serialization in HOCON format for convenient copy & paste, and Jackson can't do HOCON.

> I assume you re-introduce that library because hoplite has no render function?

Exactly. Hoplite can deserialize, but not serialize.

> Did you file a hoplite ticket to create a render function?

Nope. I did consider this myself, though, but came to the conclusion that the author of Hoplite will very likely consider this to be out of scope of his library.

Reply:
> the author of Hoplite will very likely consider this to be out of scope of his library.

Indeed he does: https://github.com/sksamuel/hoplite/issues/203#issuecomment-794471877

Reply:
> Can you explain that in the commit message?

Done.

Reply:
One last idea to avoid the dependency to config4k: IIUC, the library is now mainly used to transform an object to a Typesafe Configuration; the rendering is then done via this object. I think, for the purposes of `ConfigCommand`, this conversion could be done manually by
* using Jackson to serialize the object to a JSON string
* use the Typesafe Configuration API to load a configuration from JSON

This can probably be done in a few lines of code. Do you think this is an alternative?

Reply:
> use the Typesafe Configuration API to load a configuration from JSON

Sorry, but I still don't get how that helps to display a HOCON string to the user. And *which* Typesafe Configuration API do you mean, as strictly speaking the fact that *Hoplite* uses the Typesafe *config* library is an implementation detail, and we must not use any part of it directly without also adding it as a dependency (again).

Reply:
There are three libraries involved: _Hoplite_, Typesafe _config_, and _config4k_. ORT currently depends on the first and (transitively) the second. The rendering to HOCON is part of Typesafe _config_. In order to use it, an ORT model object has to be converted to a Typesafe _Config_; this is what _config4k_ does. I was proposing an alternative conversion which does not require an additional library.

You are right that we then need to declare Typesafe _config_ explicitly as a dependency, but at least this would not change the total set of dependencies needed by ORT.

Reply:
> but at least this would not change the total set of dependencies needed by ORT.

I must still be missing something, as *config4k* is just a thin Kotlin-wrapper around Typesafe *config*. So why does that "change the total set of dependencies"?

Reply:
Because a new dependency, _config4k_, is added.

You are right, this is only a small library and not a big deal. It is, however, a bit strange to depend on multiple libraries that do similar things (configuration management in this case).

Reply:
I've put some effort into try using *config* directly without *config4k* by doing something like

    val json = JsonMapper().registerKotlinModule().writeValueAsString(globalOptionsForSubcommands.config)
    val config = ConfigFactory.parseString(json, ConfigParseOptions.defaults().setSyntax(ConfigSyntax.JSON))
    println(config.root().render(renderOptions))

but this results in some subtle differences in the expected HOCON output (e.g. regarding type annotations), and fixing it would require adding some more code that basically just resembles what *config4k* does. So I'm giving up on this and stick to *config4k*.